python/ray/projects/projects.py

```diff
@@ -2,12 +2,103 @@
 from __future__ import division
 from __future__ import print_function
 
+import argparse
 import json
 import jsonschema
 import os
 import yaml
 
 
+class ProjectDefinition:
+    def __init__(self, current_dir):
+        """Finds .rayproject folder for current project, parse and validates it.
+
+        Args:
+            current_dir (str): Path from which to search for .rayproject.
+
+        Raises:
+            jsonschema.exceptions.ValidationError: This exception is raised
+                if the project file is not valid.
+            ValueError: This exception is raised if there are other errors in
+                the project definition (e.g. files not existing).
+        """
+        self.root = find_root(current_dir)
+        if self.root is None:
+            raise ValueError("No project root found")
+        if not self.root.endswith("/"):
+            # This is so that rsync will copy the project directory to the
+            # correct target.
+            self.root += "/"
+
+        # Parse the project YAML.
+        project_file = os.path.join(self.root, ".rayproject", "project.yaml")
+        if not os.path.exists(project_file):
+            raise ValueError("Project file {} not found".format(project_file))
+        with open(project_file) as f:
+            self.config = yaml.safe_load(f)
+
+        check_project_config(self.root, self.config)
+
+    def cluster_yaml(self):
+        """Return the project's cluster configuration filename.
+        """
+        return self.config["cluster"]
+
+    def working_directory(self):
+        """Return the project's working directory on a cluster session.
+        """
+        directory = os.path.join("~", self.config["name"])
+        if not directory.endswith("/"):
```

Comment:
nit, you can do `os.path.join("~", self.config["name"], "")`, or generally `os.path.join(path, "")` for forcing path to be a directory